## Practical 12 
## Neural Networks in Machine Learning

A neural network is a machine learning model inspired by how the human brain works. It is designed to learn patterns from data by connecting simple computing units called neurons.

Structure of a Neural Network

01) Input Layer - This is where data enters.

    Example: Age = 25, Salary = 50,000

02) Hidden Layers - This is the “brain” of the network. It performs calculations. It finds hidden patterns.

03) Output Layer - This gives the final result.

    Examples: 0 or 1 (spam or not spam), Class label (cat or dog)

### Task 01: Neural Network Implementation in Python 

#### Step 1: Import Libraries

In [1]:
import numpy as np

#### Step 2: Define the Dataset

In [2]:
# Inputs (XOR truth table) 
X = np.array([[0, 0], 
[0, 1], 
[1, 0], 
[1, 1]]) 

# Output labels 
Y = np.array([[0], 
[1], 
[1], 
[0]])

#### Step 3: Initialize Hyperparameters

Hyperparameters are settings you choose before training the model.

They control:

how big the network is,
how fast it learns, and 
how long it trains

In [3]:
input_neurons = 2        # 2 input features
hidden_neurons = 4       # can be tuned 
output_neurons = 1       # binary output 
learning_rate = 0.1 
epochs = 10000 

Input layer = 2 neurons (Because input data has 2 features) - 1 neuron for each input feature

Hidden layer = 4 neurons (These neurons learn patterns , More neurons = more learning power)

Output layer = 1 neuron (XOR output is binary → 0 or 1)

epoches - Number of times the model sees the entire dataset.

1 epoch = one full pass through all data

10000 epochs = model learns 10000 times


#### Step 4: Initialize Weights and Biases

In [ ]:
np.random.seed(1)  # reproducibility 

# Weights and biases 
wh = np.random.uniform(size=(input_neurons, hidden_neurons))  # input -> hidden 
bh = np.random.uniform(size=(1, hidden_neurons)) 

wo = np.random.uniform(size=(hidden_neurons, output_neurons))  # hidden -> output 
bo = np.random.uniform(size=(1, output_neurons))

wh (hidden weights) -This creates a weight matrix between input layer and hidden layer. Each input neuron connects to every hidden neuron.

    input_neurons = 2, hidden_neurons = 4

bh (hidden layer bias) - Each hidden neuron gets its own bias

wo (output weights) - Each hidden neuron connects to the output neuron:

    Each connection has a weight.

    hidden_neurons = 4, output_neurons = 1

bo (output bias) - The final output neuron also has a bias term.

#### Step 5: Activation Function

In [ ]:
def sigmoid(x): 
    return 1 / (1 + np.exp(-x)) 

def sigmoid_derivative(x): # calculates how much the sigmoid output changes
    return x * (1 - x)

#### Step 6: Training Loop

In [ ]:
for epoch in range(epochs): 
    # Forward Propagation 
    zh = np.dot(X, wh) + bh # calculates weighted sum for hidden layer
    ah = sigmoid(zh) # Applies sigmoid, Converts values into 0–1 range
    zo = np.dot(ah, wo) + bo # calculates weighted sum for output layer
    ao = sigmoid(zo) # prediction of the model

    # Backpropagation 
    error = Y - ao # calculates error at output layer (actual values - predicted values)
    d_output = error * sigmoid_derivative(ao) # How much output layer should change
    error_hidden = d_output.dot(wo.T) # backpropagation - which hidden neurons caused that mistake
    d_hidden = error_hidden * sigmoid_derivative(ah) # How much hidden neurons should adjust

    # Update weights and biases 
    wo += ah.T.dot(d_output) * learning_rate # Adjusts weights between hidden and output layer
    bo += np.sum(d_output, axis=0, keepdims=True) * learning_rate # Adjusts output bias
    wh += X.T.dot(d_hidden) * learning_rate # Adjusts weights from input to hidden layer
    bh += np.sum(d_hidden, axis=0, keepdims=True) * learning_rate # Adjusts hidden layer bias

    # axis = 0 means we sum across rows (for each column), keepdims=True keeps the result as a 2D array

    # Optionally print loss 
    if epoch % 1000 == 0: 
        loss = np.mean(np.square(error)) 
        print(f'Epoch {epoch}, Loss: {loss:.4f}') # Every 1000 epochs, prints error

Epoch 0, Loss: 0.3674
Epoch 1000, Loss: 0.2488
Epoch 2000, Loss: 0.2422
Epoch 3000, Loss: 0.2007
Epoch 4000, Loss: 0.1303
Epoch 5000, Loss: 0.0571
Epoch 6000, Loss: 0.0289
Epoch 7000, Loss: 0.0174
Epoch 8000, Loss: 0.0113
Epoch 9000, Loss: 0.0077


#### Step 7: Final Prediction

In [ ]:
print("\nFinal Output after Training:") 
print(np.round(ao, 3)) # Rounded output to 3 decimal places for better readability


Final Output after Training:
[[0.06 ]
 [0.938]
 [0.917]
 [0.087]]


Model made 4 predictions (so likely 4 training samples, like XOR inputs).

Each number is a probability-like output (0 to 1):

Output	Meaning

0.06	very close to 0

0.938	very close to 1

0.917	very close to 1

0.087	very close to 0

### Task 2

In [ ]:
# Neural Network from Scratch in Python 
# This tutorial covers implementation, training, and evaluation 

import numpy as np 

# ----------------------------- 
# 1. Define Activation Functions 
# ----------------------------- 
def sigmoid(x): 
    return 1 / (1 + np.exp(-x)) 

def sigmoid_derivative(x): 
    return x * (1 - x) 

# ----------------------------- 
# 2. Initialize Dataset 
# ----------------------------- 
# XOR problem 
X = np.array([[0, 0], 
[0, 1], 
[1, 0], 
[1, 1]]) 

# Output labels
Y = np.array([[0], [1], [1], [0]]) 

# -----------------------------
# 3. Initialize Network Parameters
# ----------------------------- 

#X.shape[0] - number of rows (samples) , X.shape[1] - number of columns (features)
input_layer_neurons = X.shape[1] # number of features  in data
hidden_layer_neurons = 4 
output_neurons = 1 

# Random weights and bias initialization 
np.random.seed(42) 
wh = np.random.uniform(size=(input_layer_neurons, hidden_layer_neurons)) 
bh = np.random.uniform(size=(1, hidden_layer_neurons)) 
wo = np.random.uniform(size=(hidden_layer_neurons, output_neurons)) 
bo = np.random.uniform(size=(1, output_neurons)) 

# ----------------------------- 
# 4. Train the Network 
# ----------------------------- 
learning_rate = 0.1 
epochs = 10000 

for i in range(epochs): 
    # Forward Propagation 
    hidden_layer_input = np.dot(X, wh) + bh # calculates weighted sum for hidden layer
    hidden_layer_output = sigmoid(hidden_layer_input) # Applies sigmoid, Converts values into 0–1 range
    output_layer_input = np.dot(hidden_layer_output, wo) + bo # calculates weighted sum for output layer
    predicted_output = sigmoid(output_layer_input) # prediction of the model
    
    # Backpropagation 
    error = Y - predicted_output 
    d_predicted_output = error * sigmoid_derivative(predicted_output) # How much output layer should change
    error_hidden_layer = d_predicted_output.dot(wo.T) # backpropagation - which hidden neurons caused that mistake
    d_hidden_layer = error_hidden_layer * sigmoid_derivative(hidden_layer_output) # How much hidden neurons should adjust
    
    # Update Weights and Biases 
    wo += hidden_layer_output.T.dot(d_predicted_output) * learning_rate # Adjusts weights between hidden and output layer
    bo += np.sum(d_predicted_output, axis=0, keepdims=True) * learning_rate # Adjusts output bias
    wh += X.T.dot(d_hidden_layer) * learning_rate # Adjusts weights from input to hidden layer 
    bh += np.sum(d_hidden_layer, axis=0, keepdims=True) * learning_rate # Adjusts hidden layer bias

# ----------------------------- 
# 5. Evaluate the Model 
# ----------------------------- 
print("\nFinal output after training:") 
print(np.round(predicted_output, 3))  


Final output after training:
[[0.047]
 [0.957]
 [0.925]
 [0.072]]
